File: backend/routes/quality/policies.py
from flask import Blueprint, jsonify, request
from functools import wraps
import random
from supabase_client import get_supabase

# Create blueprint
quality_policies_bp = Blueprint('quality_policies', __name__)

# Simple CORS decorator
def cors_enabled(f):
    @wraps(f)
    def decorated_function(*args, **kwargs):
        # Handle OPTIONS requests
        if request.method == 'OPTIONS':
            response = jsonify({'status': 'preflight'})
            origin = request.headers.get('Origin', 'http://localhost:3001')
            allowed_origins = ['http://localhost:3000', 'http://localhost:3001', 'http://127.0.0.1:3000', 'http://127.0.0.1:3001']
            if origin in allowed_origins:
                response.headers.add('Access-Control-Allow-Origin', origin)
            response.headers.add('Access-Control-Allow-Credentials', 'true')
            response.headers.add('Access-Control-Allow-Headers', 'Content-Type,Authorization')
            response.headers.add('Access-Control-Allow-Methods', 'GET,POST,PUT,DELETE,OPTIONS')
            return response
        
        # Handle actual requests
        response = f(*args, **kwargs)
        if hasattr(response, 'headers'):
            origin = request.headers.get('Origin', 'http://localhost:3001')
            allowed_origins = ['http://localhost:3000', 'http://localhost:3001', 'http://127.0.0.1:3000', 'http://127.0.0.1:3001']
            if origin in allowed_origins:
                response.headers.add('Access-Control-Allow-Origin', origin)
            response.headers.add('Access-Control-Allow-Credentials', 'true')
        return response
    return decorated_function

@quality_policies_bp.route('/policies', methods=['GET', 'OPTIONS'])
@cors_enabled
def get_policies():
    """Get quality policies from Supabase"""
    try:
        supabase = get_supabase()

        # Get pagination parameters
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))
        offset = (page - 1) * limit

        # Get filter parameters
        search = request.args.get('search', '')
        category_filter = request.args.get('category', '')
        status_filter = request.args.get('status', '')

        # Build query
        query = supabase.table('quality_policy').select('*', count='exact')

        # Apply filters
        if search:
            query = query.or_(f"policy_name.ilike.%{search}%,department.ilike.%{search}%,responsible_person.ilike.%{search}%")
        if category_filter:
            query = query.eq('category', category_filter)
        if status_filter:
            query = query.eq('compliance_status', status_filter)

        # Get total count for pagination
        total_query = query
        total_result = supabase.table('quality_policy').select('*', count='exact').execute()
        total_count = total_result.count if hasattr(total_result, 'count') else 0

        # Apply pagination
        query = query.range(offset, offset + limit - 1)

        # Execute query
        result = query.execute()

        if not hasattr(result, 'data'):
            return jsonify({
                'success': False,
                'error': 'Failed to fetch data from Supabase'
            }), 500

        # Map database fields to frontend expected fields
        policies = []
        for policy in result.data:
            mapped_policy = {
                'id': policy.get('policy_id'),
                'title': policy.get('policy_name', ''),
                'description': policy.get('description', ''),
                'category': policy.get('category', 'General'),
                'department': policy.get('department', ''),
                'compliance_status': policy.get('compliance_status', 'pending'),
                'compliance_score': policy.get('compliance_score', 0),
                'next_review_date': policy.get('next_due_date', ''),
                'responsible_person': policy.get('responsible_person', '')
            }
            policies.append(mapped_policy)

        total_pages = (total_count + limit - 1) // limit if total_count > 0 else 1

        return jsonify({
            'success': True,
            'data': policies,
            'pagination': {
                'currentPage': page,
                'totalPages': total_pages,
                'totalItems': total_count,
                'limit': limit
            }
        })

    except Exception as e:
        print(f"Error fetching policies: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_policies_bp.route('/policies/analytics', methods=['GET', 'OPTIONS'])
@cors_enabled
def get_policies_analytics():
    """Get policy analytics"""
    try:
        supabase = get_supabase()

        # Get all policies for analytics
        result = supabase.table('quality_policy').select('*').execute()

        if not result.data:
            # Return mock analytics if no data
            analytics = {
                'compliance_trends': [
                    {'month': 'Jan', 'rate': 80},
                    {'month': 'Feb', 'rate': 82},
                    {'month': 'Mar', 'rate': 85},
                    {'month': 'Apr', 'rate': 87},
                    {'month': 'May', 'rate': 90},
                    {'month': 'Jun', 'rate': 92}
                ],
                'upcoming_deadlines': [],
                'policy_compliance': []
            }
        else:
            # Calculate real analytics from data
            from datetime import datetime, timedelta
            import random

            # Compliance trends (mock for now, could be calculated from historical data)
            compliance_trends = [
                {'month': 'Jan', 'rate': 80},
                {'month': 'Feb', 'rate': 82},
                {'month': 'Mar', 'rate': 85},
                {'month': 'Apr', 'rate': 87},
                {'month': 'May', 'rate': 90},
                {'month': 'Jun', 'rate': 92}
            ]

            # Upcoming deadlines - get policies with next_due_date within 60 days
            upcoming_deadlines = []
            today = datetime.now().date()

            for policy in result.data[:10]:  # Limit to first 10 for display
                try:
                    due_date = datetime.strptime(policy['next_due_date'], '%Y-%m-%d').date()
                    days_left = (due_date - today).days
                    if 0 <= days_left <= 60:  # Only show upcoming or slightly overdue
                        upcoming_deadlines.append({
                            'policy': policy['policy_name'],
                            'days_left': max(0, days_left)
                        })
                except (ValueError, KeyError):
                    continue

            # Sort by days left
            upcoming_deadlines.sort(key=lambda x: x['days_left'])

            # Policy compliance status
            policy_compliance = []
            for policy in result.data[:10]:  # Limit to first 10
                policy_compliance.append({
                    'policy': policy['policy_name'],
                    'status': policy['compliance_status']
                })

            analytics = {
                'compliance_trends': compliance_trends,
                'upcoming_deadlines': upcoming_deadlines,
                'policy_compliance': policy_compliance
            }

        return jsonify({
            'success': True,
            'data': analytics
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

File: backend/controllers/quality/policies.js
const { supabase } = require('../../supabase_client');

// SUB-MODULE 5: POLICY COMPLIANCE STATUS
exports.getAllPolicies = async (req, res) => {
  try {
    const { page = 1, limit = 10, department, compliance_status } = req.query;
    
    let query = supabase
      .from('policies')
      .select('*', { count: 'exact' })
      .order('next_review_date', { ascending: true });

    if (department) query = query.eq('department', department);
    if (compliance_status) query = query.eq('compliance_status', compliance_status);

    const from = (page - 1) * limit;
    const to = from + limit - 1;

    const { data, error, count } = await query.range(from, to);

    if (error) throw error;

    res.json({
      success: true,
      data: data || [],
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: count || 0,
        totalPages: Math.ceil((count || 0) / limit)
      }
    });
  } catch (error) {
    console.error('Error fetching policies:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch policy data'
    });
  }
};

exports.addPolicy = async (req, res) => {
  try {
    const policyData = {
      ...req.body,
      compliance_status: 'pending_review',
      last_reviewed: new Date().toISOString(),
      compliance_score: req.body.compliance_score || 0,
      documents: req.body.documents || []
    };

    const { data, error } = await supabase
      .from('policies')
      .insert([policyData])
      .select();

    if (error) throw error;

    res.status(201).json({
      success: true,
      data: data[0],
      message: 'Policy added successfully'
    });
  } catch (error) {
    console.error('Error adding policy:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to add policy'
    });
  }
};

exports.updatePolicy = async (req, res) => {
  try {
    const { id } = req.params;
    
    const { data, error } = await supabase
      .from('policies')
      .update(req.body)
      .eq('id', id)
      .select();

    if (error) throw error;

    if (!data || data.length === 0) {
      return res.status(404).json({
        success: false,
        error: 'Policy not found'
      });
    }

    res.json({
      success: true,
      data: data[0],
      message: 'Policy updated successfully'
    });
  } catch (error) {
    console.error('Error updating policy:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update policy'
    });
  }
};

exports.deletePolicy = async (req, res) => {
  try {
    const { id } = req.params;
    
    const { error } = await supabase
      .from('policies')
      .delete()
      .eq('id', id);

    if (error) throw error;

    res.json({
      success: true,
      message: 'Policy deleted successfully'
    });
  } catch (error) {
    console.error('Error deleting policy:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete policy'
    });
  }
};

exports.getNonCompliantPolicies = async (req, res) => {
  try {
    const { data, error } = await supabase
      .from('policies')
      .select('*')
      .eq('compliance_status', 'non_compliant')
      .order('due_date', { ascending: true });

    if (error) throw error;

    res.json({
      success: true,
      data: data || []
    });
  } catch (error) {
    console.error('Error fetching non-compliant policies:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch non-compliant policies'
    });
  }
};

exports.getPoliciesDueForReview = async (req, res) => {
  try {
    const today = new Date().toISOString().split('T')[0];
    
    const { data, error } = await supabase
      .from('policies')
      .select('*')
      .lte('next_review_date', today)
      .order('next_review_date', { ascending: true });

    if (error) throw error;

    res.json({
      success: true,
      data: data || []
    });
  } catch (error) {
    console.error('Error fetching policies due for review:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch policies due for review'
    });
  }
};

exports.getPolicyAnalytics = async (req, res) => {
  try {
    // Compliance trends
    const complianceTrends = [
      { month: 'Jan', rate: 80 },
      { month: 'Feb', rate: 82 },
      { month: 'Mar', rate: 85 },
      { month: 'Apr', rate: 87 },
      { month: 'May', rate: 90 },
      { month: 'Jun', rate: 92 }
    ];

    // Policy-wise compliance status
    const { data: policyData } = await supabase
      .from('policies')
      .select('title, compliance_status');

    const policyCompliance = policyData ? 
      policyData.map(policy => ({
        policy: policy.title,
        status: policy.compliance_status
      })) : [];

    // Upcoming deadlines
    const today = new Date();
    const thirtyDaysFromNow = new Date(today.getTime() + 30 * 24 * 60 * 60 * 1000);

    const { data: deadlineData } = await supabase
      .from('policies')
      .select('title, next_review_date')
      .gte('next_review_date', today.toISOString())
      .lte('next_review_date', thirtyDaysFromNow.toISOString())
      .order('next_review_date', { ascending: true });

    const upcomingDeadlines = deadlineData ? 
      deadlineData.map(policy => {
        const daysLeft = Math.ceil((new Date(policy.next_review_date) - today) / (1000 * 60 * 60 * 24));
        return {
          policy: policy.title,
          days_left: daysLeft
        };
      }) : [];

    const analytics = {
      compliance_trends: complianceTrends,
      policy_compliance: policyCompliance,
      upcoming_deadlines: upcomingDeadlines
    };

    res.json({
      success: true,
      data: analytics
    });
  } catch (error) {
    console.error('Error fetching policy analytics:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch policy analytics'
    });
  }
};
